data/sensors_recent.php

```diff
@@ -138,63 +142,104 @@ function node_button($id) {
+    <a class="dropdown-item gw_filter"   href="$filter_url_html">Show only messages received by $gw_html</a>
+    <a class="dropdown-item gw_location" href="$location_url_html">Show gateway on map</a>
+    <a class="dropdown-item gw_loc_ttn"  href="$location_ttn_html">Show gateway on TTN</a>
   </div>
 </div>
 EOF;
 	}
 
+
 	function output_cell($rowspan, $data) {
 		echo("  <td rowspan=\"$rowspan\"> " . $data . "</td>\n");
 	}
 
+
+	// main loop:
 	while($row = $result->fetch_array(MYSQLI_ASSOC)) {
 		if ($row['message']) {
 			$message = json_decode($row['message'], true);
-			$metadata = $message['metadata'];
 
-			if (!array_key_exists('gateways', $metadata)) {
-				$gateways = [];
+			if ($row['source'] == 'ttn.v1') {
+				$metadata = $message['metadata'];
+				$metadata['counter'] = $message['counter'];
 
-				// Convert TTN staging to production format
+				$gateways = [];
+	
+				// Convert TTN v1=staging to v2=production format
 				foreach ($metadata as $meta) {
 					$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
 					$meta['snr'] = $meta['lsnr'];
 					$gateways[] = $meta;
 				}
-
+	
 				$metadata = [
-					'gateways' => $gateways,
-					'frequency' => $metadata[0]['frequency'],
-					'data_rate' => $metadata[0]['datarate'],
-					'coding_rate' => $metadata[0]['codingrate'],
+					'gateways'    => $message['metadata']['gateways'],
+					'frequency'   => $message['metadata']['frequency'],
+					'data_rate'   => $message['metadata']['datarate'],
+					'coding_rate' => $message['metadata']['codingrate'],
+					'counter'     => $message['counter'],
 				];
-			}
 
+			} else if ($row['source'] == 'ttn.v2') {
+				$metadata = $message['metadata'];
+				$metadata['counter'] = $message['counter'];
+
+			} else if ($row['source'] == 'ttn.v3') {
+				$metadata = $message['uplink_message']['rx_metadata'];
+	
+				$gateways = [];
+
+				// Convert TTN v3 to v2 format
+				foreach ($metadata as $meta) {
+					$meta['gtw_id'] = $meta['gateway_ids']['gateway_id'];
+					$meta['gtw_eui'] = "";
+					if (isset($meta['gateway_ids']['eui']))
+						$meta['gtw_eui'] = $meta['gateway_ids']['eui'];
+					if ($meta['gtw_id'] == 'packetbroker')
+						$meta['gtw_id'] = 'unknown';
+					if (!isset($meta['snr'])) 
+						$meta['snr'] = $null_snr;
+
+					$gateways[] = $meta;
+				}
+	
+				$spreading_factor = $message['uplink_message']['settings']['data_rate']['lora']['spreading_factor'];
+				$bandwidth =        $message['uplink_message']['settings']['data_rate']['lora']['bandwidth'] / 1000;
+				$data_rate = sprintf("SF%sBW%s", $spreading_factor, $bandwidth);
+				$metadata = [
+					'gateways'    => $gateways,
+					'frequency'   => $message['uplink_message']['settings']['frequency'] / 1000000.,
+					'coding_rate' => $message['uplink_message']['settings']['coding_rate'],
+					'data_rate'   => $data_rate,
+					'counter'     => $message['uplink_message']['f_cnt'],
```

Comment:
Turns out that in v3, a `f_cnt` of 0 is currently being emitted from the generated JSON (see https://github.com/TheThingsNetwork/lorawan-stack/issues/3874), so we should handle that here. I would suggest making this change:
```suggestion
					'counter'     => isset($message['uplink_message']['f_cnt']) ? $message['uplink_message']['f_cnt'] : 0,
```

Reply:
This can be seen in the messags from 2022 in the past evening, I've been resetting it a lot:

![image](https://user-images.githubusercontent.com/194491/119060435-7315a080-b9d2-11eb-8ae7-8327c24d68f2.png)